### PR TITLE
Web server in k8sapp initiative repo

### PR DESCRIPTION
**Pull request by didierofrivia**
Part of https://github.com/3scale/k8sapp-initiative/pull/16

Building the image:

```bash
cd web-server
DOCKER_BUILDKIT=1 docker build -f Dockerfile -t [REPO] .
```
Then replace the `web-server-deployment.yml` image for the one you built.

It's currently deployed on http://k8sinitiative.apps.dev-eng-ocp4-5.dev.3sca.net/products

### CODE REVIEW

**Review by damianpm — CHANGES_REQUESTED** [review]
Please merge https://github.com/3scale/k8sapp-initiative/pull/16 or change base branch, as `web-server_deployment.yml` is not in `master` yet

File: web-server/pkg/main.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func handleProductsPage(w http.ResponseWriter, r *http.Request) {
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	v1alpha1.AddToScheme(scheme.Scheme)
```
Comment (eguzki) [inline web-server/pkg/main.go]: adding v1alpha1 types at request time?  why not 
```
func init() {
    v1alpha1.AddToScheme(scheme.Scheme)
}
```

File: web-server/pkg/apis/k8sinitiative.3scale.net/v1alpha1/types.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+//go:generate controller-gen object paths=$GOFILE
+
+type Product struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              ProductSpec `json:"spec"`
+}
+
+type ProductSpec struct {
```
Comment (eguzki) [inline web-server/pkg/apis/k8sinitiative.3scale.net/v1alpha1/types.go:13]: All attributes are required?

My advise, add all of them as optional. It is much easier for latter deprecation/updating...

´´´
Description   *string `json:"description,omitempty"`
´´´

File: web-server/pkg/main.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+	crdConfig := *config
+	crdConfig.ContentConfig.GroupVersion = &schema.GroupVersion{Group: v1alpha1.GroupName, Version: v1alpha1.GroupVersion}
+	crdConfig.APIPath = "/apis"
+	crdConfig.NegotiatedSerializer = serializer.NewCodecFactory(scheme.Scheme)
+	crdConfig.UserAgent = rest.DefaultKubernetesUserAgent()
+
+	threescaleRestClient, err := rest.RESTClientFor(&crdConfig)
```
Comment (eguzki) [inline web-server/pkg/main.go]: a) this is not a threescale rest client, but a k8s rest client. 

b) not required to do at request time. Maybe at initialization stage?

c) Why use such low level rest client? There are higher level clients

File: web-server/pkg/main.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		panic(err)
+	}
+	result := v1alpha1.ProductList{}
+
+	getErr := threescaleRestClient.
```
Comment (eguzki) [inline web-server/pkg/main.go]: what if getErr is not nil?

File: web-server/pkg/main.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		Get().
+		Namespace("k8sinitiative").
+		Resource("products").
+		Do().
+		Into(&result)
+
+	fmt.Printf("%d results found: %+v\n", len(result.Items), result)
```
Comment (eguzki) [inline web-server/pkg/main.go:97]: For generic print func, I use:
```
import (                                                                         
    "encoding/json"                                                              
    "fmt"                                                                        
                                                                                 
    "github.com/davecgh/go-spew/spew"                                            
)                                                                                
                                                                                 
func Print(obj interface{}) {                                                    
    fmt.Println("######################################################")        
    fmt.Println("# Dump")                                                        
    fmt.Println("######################################################")        
    spew.Dump(obj)                                                               
    fmt.Println("######################################################")        
    fmt.Println("# JSON ")                                                       
    fmt.Println("######################################################")        
    jsonData, err := json.MarshalIndent(obj, "", "  ")                           
    if err != nil {                                                              
        panic(err)                                                               
    }                                                                            
    fmt.Println(string(jsonData))                                                
}                                                                                

```

File: web-server/pkg/main.go (hunk 3 above)
Comment (didierofrivia) [reply web-server/pkg/main.go]: Answering C: Mainly for educational purposes :)

File: web-server/pkg/main.go (hunk 5 above)
Comment (didierofrivia) [reply web-server/pkg/main.go:97]: Will treasure this for future logs

File: web-server/pkg/apis/k8sinitiative.3scale.net/v1alpha1/types.go [hunk 6]
```diff
@@ -11,12 +11,12 @@ type Product struct {
 }
 
 type ProductSpec struct {
-	ID            int    `json:"id"`
-	Description   string `json:"description"`
-	AppsCount     int    `json:"apps_count"`
-	BackendsCount int    `json:"backends_count"`
-	Name          string `json:"name"`
-	SystemName    string `json:"system_name"`
+	ID            *int    `json:"id"`
```
Comment (eguzki) [inline web-server/pkg/apis/k8sinitiative.3scale.net/v1alpha1/types.go]: Should be 
```
ID            *int    `json:"id,omitempty"`
```

**Review by damianpm — APPROVED** [review]

**Pull request merged**